job-runner/chain/traverser.go

```diff
@@ -534,41 +565,90 @@ func (t *traverser) shutdown() {
 	// stopping the suspended reaper.
 	select {
 	case <-suspendedReaperChan:
-	case <-time.After(t.stopTimeout):
+	case <-timeout:
 		t.logger.Warnf("timed out waiting for jobs to stop - stopping reaper")
 		t.reaper.Stop()
 	}
 	close(t.doneChan)
 }
 
 // stopRunningJobs stops all currently running jobs.
-func (t *traverser) stopRunningJobs() error {
+func (t *traverser) stopRunningJobs(timeout <-chan time.Time) error {
+	// To stop all running jobs without race coditions, we need to know:
+	//   1. runJobs is done, won't start any more goroutines
+	//   2. All in-flight runJob goroutines have added themselves to runner repo
+	// First is easy: wait for it to close pendingChan. Second is like a wait
+	// group wait: runJobs add +1 to pending when goroutine starts, and -1 after
+	// it adds itself to runner repo. So all runJob goroutines have added
+	// themselves to the runner repo when pending == 0.
+	//
+	// The shutdown sequence is:
+	//   1. close(stopChan): runJob goroutines (RGs) don't run if closed. It's
+	//      as if the job never ran. This allows runJobChan to drain and prevents
+	//      runnerRepo from blocking because the chan is unbuffered. This is done
+	//      in the for loop, before launching the goroutine, so that a closed
+	//      pendingChan (step 4) guarantees that runJobs either didn't run an
+	//      RG or it did and added it to pending count (because the loop won't
+	//      exit until running pre-goroutine code, and pendingChan is only closed
+	//      after loop exits).
+	//   2. Stop runningReaper: This stops new/next jobs into runJobChan, which
+	//      is being drained because of step 1.
+	//   3. close(runJobChan): When runningReaper.Run returns, the goroutine in
+	//      in traverser.Run closes runJobChan. Since runningReaper is only thing
+	//      that sends to runJobChan, it must be closed like this so runningReaper
+	//      doesn't panic on "send on closed channel".
+	//   4. close(pendingChan): Given step 3 and step 1, eventually runJobChan
+	//      will drain and runJobs() will return, closing pendingChan when it does.
+	//   5. Call stopRunningJobs: This func waits for step 4, which ensures no
+	//      more RGs. And given step 1, we're assured that all in-flight RGs
+	//      have added themsevs to pending count. Therefore, this func waits for
+	//      pending count == 0 which means all RGs have added themselves to the
+	//      runner repo.
+	//   6. Stop all active runners in runner repo.
+
+	// Wait for runJobs to return
+	select {
+	case <-t.pendingChan:
+	case <-timeout:
+		return fmt.Errorf("stopRunningJobs: timeout waiting for pendingChan")
+	}
+
+	// Wait for in-flight runJob goroutines to add themselves to runner repo
+	if n := atomic.LoadInt64(&t.pending); n > 0 {
+		for atomic.LoadInt64(&t.pending) > 0 {
+			select {
+			case <-timeout:
+				return fmt.Errorf("stopRunningJobs: timeout waiting for pending count")
+			default:
+				time.Sleep(100 * time.Millisecond)
+			}
```

Comment:
The nit-iest nit: In cases like this I generally prefer
```
select {
case <-timeout:
    return fmt.Errorf("stopRunningJobs: timeout waiting for pending count")
case <-time.After(100 * time.Millisecond):
}
```
so the timeout can still happen during the wait. It matters less when the wait is on the order of milliseconds though.

Reply:
Normally yes but with a 100ms sleep, we won't wait know any longer than the timeout as perceived by a human.